File: comp_translate/SaveLoadTxt.jl
module SaveLoadTxt
export dump_txt, load_txt

const HEADER = "# translate2compare 1.0"

import Printf: @sprintf

# ── helpers ────────────────────────────────────────────────────────────────
function _describe(obj)
    if obj isa Integer
        return "scalar", "int64", nothing
    elseif obj isa AbstractFloat
        obj isa Float64 || throw(ArgumentError("Only Float64 supported"))
        return "scalar", "float64", nothing
    elseif obj isa AbstractVector
        T = eltype(obj)
        dtype = T <: Integer ? "int64" :
                T == Float64  ? "float64" :
                throw(ArgumentError("Unsupported eltype $T"))
        return "vector", dtype, (length(obj),)
    elseif obj isa AbstractMatrix
        T = eltype(obj)
        dtype = T <: Integer ? "int64" :
                T == Float64  ? "float64" :
                throw(ArgumentError("Unsupported eltype $T"))
        return "matrix", dtype, size(obj)
    else
        throw(ArgumentError("Unsupported object $(typeof(obj))"))
    end
end

_token(x::Int)      = string(x)
_token(x::Float64)  = @sprintf("%.17g", x)

function _write_object(io, obj)
    kind, dtype, dims = _describe(obj)
    if kind == "scalar"
        println(io, "# object $kind $dtype")
        println(io, _token(obj))
    elseif kind == "vector"
        n = dims[1]; println(io, "# object $kind $dtype $n")
        println(io, join((_token(x) for x in obj), ' '))
    else  # matrix
        r, c = dims; println(io, "# object $kind $dtype $r $c")
        for i in 1:r
            println(io, join((_token(obj[i,j]) for j in 1:c), ' '))
        end
    end
end

"""
    dump_txt(path, objects; append=true)

Serialise each object in `objects` to `path` (translate2compare format).
Set `append=false` to overwrite the file, otherwise new objects are appended.
"""
function dump_txt(path::AbstractString, objects; append::Bool=true)
    isempty(objects) && return
    mode = append ? "a" : "w"
    first_write = !isfile(path) || !append
    open(path, mode) do io
        first_write && println(io, HEADER)
        for obj in objects
            _write_object(io, obj)
        end
    end
    return nothing
end

# ── reading ────────────────────────────────────────────────────────────────
function _parse_header(io)
    hdr = readline(io) |> strip
    hdr == HEADER || error("Bad header (got “$(hdr)”)")
end

function _parse_descriptor(line)
    startswith(line, "# object ") || error("Malformed descriptor")
    parts = split(strip(line[10:end]))
    length(parts) ≥ 2 || error("Descriptor too short")
    # kind, dtype = parts[1:2]...
    # dims = length(parts) > 2 ? parse.(Int, parts[3:end]) : Int[]
    # return kind, dtype, dims
    kind = parts[1]
    dtype = parts[2]
    dims = length(parts) > 2 ? parse.(Int, parts[3:end]) : Int[]
    return kind, dtype, dims
end

function _readtokens(io, n)
    toks = split(readline(io))
    length(toks) == n || error("Expected $n tokens")
    return toks
end

function _convert(toks, dtype)
    if dtype == "int64"
        return parse.(Int64, toks)
    elseif dtype == "float64"
        return parse.(Float64, toks)
    else
        error("Unsupported dtype $dtype")
    end
end

"""
    load_txt(path) → Vector{Any}

Read every object stored in `path` and return them in file order.
"""
function load_txt(path::AbstractString)
    objs = Any[]
    open(path) do io
        _parse_header(io)
        while !eof(io)
            line = strip(readline(io))
            isempty(line) && continue
            k, dt, dims = _parse_descriptor(line)
            if k == "scalar"
                push!(objs, dt == "int64" ? parse(Int64, strip(readline(io)))
                                          : parse(Float64, strip(readline(io))))
            elseif k == "vector"
                n = (length(dims) == 1) ? dims[1] : error("Vector needs 1 dim")
                push!(objs, _convert(_readtokens(io, n), dt))
            elseif k == "matrix"
                r, c = (length(dims) == 2) ? dims : error("Matrix needs 2 dims")
                M = dt == "int64" ? Array{Int64}(undef, r, c) : Array{Float64}(undef, r, c)
                for i in 1:r
                    M[i, :] = _convert(_readtokens(io, c), dt)
                end
                push!(objs, M)
            else
                error("Unknown kind $k")
            end
        end
    end
    return objs
end

end # module

File: comp_translate/SaveLoadTxt.py
"""
save_load_txt.py
================
Plain-text serializer/deserializer for numeric scalars, 1-D vectors and
2-D matrices in the **translate2compare 1.0** format.

API
---
dump_txt(path, objects, append=True)  -> None
load_txt(path)                        -> list
"""
from __future__ import annotations

import os
from pathlib import Path
from typing import Sequence, List, Tuple, Union, Any

import numpy as np

HEADER = "# translate2compare 1.0"
Scalar = Union[int, float]
Vector = np.ndarray  # ndim == 1
Matrix = np.ndarray  # ndim == 2
Obj = Union[Scalar, Vector, Matrix]


# ─────────────────────────── helpers ────────────────────────────
def _token(x: Scalar) -> str:
    """Return an ASCII token with full precision for Float64."""
    return format(x, ".17g") if isinstance(x, float) else str(int(x))


def _describe(obj: Obj) -> Tuple[str, str, Tuple[int, ...]]:
    """
    kind ∈ {scalar, vector, matrix}
    dtype ∈ {int64, float64}
    dims  = () | (N,) | (rows, cols)
    """
    if np.isscalar(obj):
        if isinstance(obj, (int, np.integer)):
            return "scalar", "int64", ()
        elif isinstance(obj, (float, np.floating)):
            return "scalar", "float64", ()
        raise ValueError(f"Unsupported scalar type {type(obj)}")

    if not isinstance(obj, np.ndarray):
        raise ValueError(f"Unsupported object {type(obj)}")

    if obj.dtype == np.int64:
        dtype = "int64"
    elif obj.dtype == np.float64:
        dtype = "float64"
    else:
        raise ValueError(f"Unsupported dtype {obj.dtype}")

    if obj.ndim == 1:
        return "vector", dtype, (obj.shape[0],)
    elif obj.ndim == 2:
        return "matrix", dtype, obj.shape
    raise ValueError("Only 1-D or 2-D arrays supported")


def _write_object(fh, obj: Obj) -> None:
    kind, dtype, dims = _describe(obj)
    if kind == "scalar":
        fh.write(f"# object {kind} {dtype}\n")
        fh.write(_token(obj) + "\n")

    elif kind == "vector":
        n, = dims
        fh.write(f"# object {kind} {dtype} {n}\n")
        fh.write(" ".join(_token(x) for x in obj) + "\n")

    elif kind == "matrix":
        rows, cols = dims
        fh.write(f"# object {kind} {dtype} {rows} {cols}\n")
        for r in range(rows):
            fh.write(" ".join(_token(x) for x in obj[r, :]) + "\n")


def dump_txt(path: str | Path,
             objects: Sequence[Obj],
             append: bool = True) -> None:
    """
    Write every object in *objects* to *path* (translate2compare format).

    Parameters
    ----------
    path : str or Path
    objects : iterable of scalars / numpy vectors / numpy matrices
    append : bool, default True
        If False, the file is overwritten; if True, new blocks are appended.
    """
    if not objects:
        return

    path = Path(path)
    mode = "a" if append else "w"
    header_needed = not (append and path.exists())

    with path.open(mode, encoding="utf-8") as fh:
        if header_needed:
            fh.write(HEADER + "\n")
        for obj in objects:
            _write_object(fh, obj)


# ────────────────────────── reading side ─────────────────────────
def _expect_header(fh):
    line = fh.readline()
    if not line:
        raise ValueError("Empty file")
    if line.rstrip("\n") != HEADER:
        raise ValueError(f"Bad header: {line.strip()!r}")


def _parse_descriptor(line: str):
    if not line.startswith("# object "):
        raise ValueError(f"Malformed descriptor line: {line!r}")
    parts = line[9:].strip().split()
    if len(parts) < 2:
        raise ValueError(f"Incomplete descriptor: {line!r}")
    kind, dtype, *dims = parts
    dims = tuple(int(d) for d in dims)
    return kind, dtype, dims


def _convert(tokens: List[str], dtype: str):
    return (np.array(tokens, dtype=np.int64) if dtype == "int64"
            else np.array(tokens, dtype=np.float64))


def load_txt(path: str | Path) -> List[Obj]:
    """
    Parse *path* and return every stored object in order.

    Scalars become int / float. Vectors & matrices become NumPy arrays.
    """
    objs: List[Obj] = []
    path = Path(path)

    with path.open("r", encoding="utf-8") as fh:
        _expect_header(fh)

        while True:
            pos = fh.tell()
            line = fh.readline()
            if not line:          # EOF
                break
            line = line.strip()
            if not line:          # skip blanks
                continue

            # rewind so descriptor parser sees full line
            fh.seek(pos)
            desc = fh.readline().rstrip("\n")
            kind, dtype, dims = _parse_descriptor(desc)

            if kind == "scalar":
                tok = fh.readline().strip()
                val: Obj = int(tok) if dtype == "int64" else float(tok)
                objs.append(val)

            elif kind == "vector":
                (n,) = dims
                toks = fh.readline().split()
                if len(toks) != n:
                    raise ValueError("Token count mismatch in vector")
                objs.append(_convert(toks, dtype))

            elif kind == "matrix":
                rows, cols = dims
                rows_data = [fh.readline().split() for _ in range(rows)]
                if any(len(r) != cols for r in rows_data):
                    raise ValueError("Token count mismatch in matrix")
                mat = _convert([tok for row in rows_data for tok in row],
                               dtype).reshape(rows, cols)
                objs.append(mat)

            else:
                raise ValueError(f"Unknown kind {kind!r}")

    return objs
